File: backend/pkg/serving/providers/mock.py
from __future__ import annotations

from pkg.apis.inference.models import InferenceRequest, InferenceResponse


class MockLLMProvider:
    """Deterministic local provider. No GPU required."""

    name = "mock"

    def __init__(self) -> None:
        self.calls: list[InferenceRequest] = []

    async def infer(self, request: InferenceRequest) -> InferenceResponse:
        self.calls.append(request)
        preview = request.prompt.strip().replace("\n", " ")
        if len(preview) > 80:
            preview = preview[:77] + "..."
        output = (
            f"[mock-llm] model={request.model} "
            f"echo={preview!r} tokens={request.max_tokens}"
        )
        return InferenceResponse(
            model=request.model,
            output=output,
            finish_reason="stop",
            usage={"prompt_tokens": max(1, len(request.prompt) // 4), "completion_tokens": 16},
            metadata={"provider": self.name},
        )

File: backend/pkg/serving/providers/mock.go
package providers

import (
	"context"
	"fmt"
	"sync"

	"github.com/google/uuid"
	"github.com/spider/spider/pkg/apis"
)

type LLMProvider interface {
	Name() string
	Infer(ctx context.Context, request apis.InferenceRequest) (apis.InferenceResponse, error)
}

type MockLLMProvider struct {
	mu    sync.Mutex
	Calls []apis.InferenceRequest
}

func NewMockLLMProvider() *MockLLMProvider {
	return &MockLLMProvider{Calls: make([]apis.InferenceRequest, 0)}
}

func (m *MockLLMProvider) Name() string { return "mock" }

func (m *MockLLMProvider) Infer(ctx context.Context, request apis.InferenceRequest) (apis.InferenceResponse, error) {
	_ = ctx
	m.mu.Lock()
	m.Calls = append(m.Calls, request)
	m.mu.Unlock()

	preview := request.Prompt
	if len(preview) > 120 {
		preview = preview[:120] + "..."
	}
	output := fmt.Sprintf("[mock-llm:%s] %s", request.Model, preview)
	finish := "stop"
	tokens := len(request.Prompt)/4 + 32
	return apis.InferenceResponse{
		RequestID:    uuid.New().String(),
		Model:        request.Model,
		Output:       &output,
		FinishReason: &finish,
		Usage: map[string]int{
			"prompt_tokens":     len(request.Prompt) / 4,
			"completion_tokens": 32,
			"total_tokens":      tokens,
		},
		Metadata: map[string]interface{}{"provider": "mock"},
	}, nil
}

func (m *MockLLMProvider) CallCount() int {
	m.mu.Lock()
	defer m.mu.Unlock()
	return len(m.Calls)
}

var Registry = map[string]func() LLMProvider{
	"mock": func() LLMProvider { return NewMockLLMProvider() },
}
